Decrement CIDR refcount on rule replace

raybejjani opened the pull request:
When a rule replaces another the CIDRs in ToCIDR and ToCIDRSet may
change. We must release these prefixes in order to keep the refcount,
used to manage the corresponding identities, correct.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5877)
<!-- Reviewable:end -->

joestringer commented on daemon/policy.go:
```diff
@@ -207,20 +208,38 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
+			metrics.PolicyImportErrors.Inc()
+			log.WithError(err).WithField("prefixes", prefixes).Warn(
+				"Failed to release CIDRs in replaced rules during policy add with replace")
+			return d.policy.GetRevision(), err
```
If we're going to error out in this case, then I think we also need to `ipcache.ReleaseCIDRs(prefixes)` which is allocated immediately above this opts != nil case and assumes there are no subsequent failures.

I also note that there is a `d.Policy.Mutex.Lock()` immediately above this, but presumably from within this code we could do a release of CIDRs through to the KVstore and potentially hold that lock for a very long time. Do you think that there's something we can do about that?

raybejjani replied:
Good points both. I moved the new release handling to after the the section where we hold the policy repo lock. I don't know why I thought it was important to hold it for both operations, but looking at it again I think it's ok to do the releasing later. We could actually defer this but I thought that would create bugs.

raybejjani committed:
policy: Decrement CIDR refcount on rule replace

When a rule replaces another the CIDRs in ToCIDR and ToCIDRSet may
change. We must release these prefixes in order to keep the refcount,
used to manage the corresponding identities, correct.

Signed-off-by: Ray Bejjani <ray@covalent.io>

joestringer commented on daemon/policy.go:
```diff
@@ -206,27 +207,56 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
+		if err := ipcache.ReleaseCIDRs(removedPrefixes); err != nil {
+			_ = d.prefixLengths.Delete(prefixes)
+			metrics.PolicyImportErrors.Inc()
+			log.WithError(err).WithField("prefixes", prefixes).Warn(
```
```suggestion
			log.WithError(err).WithField("prefixes", removedPrefixes).Warn(
```

joestringer commented on daemon/policy.go:
```diff
@@ -206,27 +207,56 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
+				log.WithError(errReleaseCIDR).WithField("prefixes", prefixes).Warn(
+					"Failed to release CIDRs allocated for new rules during policy add with replace error cleanup")
+			}
+			return d.policy.GetRevision(), err
```
This still looks a bit uneasy to me. All of the other error cases result in no changes being made to the policy repository.

This case, on the other hand, successfully changes the policy repository but then backs out any ipcache propagation and also returns an error. So we no longer know what the policy enforcement will be based upon the result of the policy add.

I suspect this is the motivation behind your original version keeping the lock across this: You could find the rules to remove, release the corresponding CIDRs, then (in theory) if anything fails, back out. If instead you look up the rules to delete, release the lock, free the CIDRs, then take the lock again and remove the rules, it's possible for the repository to change in between, which means that there may be new CIDRs that are not yet released. :/

From a k8s perspective, maybe the current proposed behaviour is OK: By returning an error, there will be an attempt to reinject the policy again which at least provides an opportunity to retry these failed operations...

I'll be looking at https://github.com/cilium/cilium/issues/5686 for the 1.4 cycle which should hopefully simplify this case in the longer term, however that won't be any good for a backport. So I guess the question is what the best compromise would be for handling this case for existing releases - and maybe this is it. I'd welcome additional eyes on this though.

raybejjani replied:
This is a good point. Even if a re-insert happened because of the error, we have already leaked the release. I've made this a log that doesn't return, and matched the operation order to PolicyDelete.
Nothing will crash immediately but it will leak. I'm not sure how else to do this without refactoring all releases into an async scheme (as tgraf is implying). That would mean changing PolicyDelete and the ipcache implementation, at least, and perhaps any other places we want to release identities etc.

joestringer reviewed (changes requested):
There's one log message that looks like it needs updating, and I have another comment as well..

joestringer commented on daemon/policy.go:
```diff
@@ -206,27 +207,47 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
+		log.WithField("prefixes", removedPrefixes).Debug("Decrementing replaced CIDR refcounts when adding rules")
+		if err := ipcache.ReleaseCIDRs(removedPrefixes); err != nil {
+			log.WithError(err).WithField("prefixes", removedPrefixes).Error(
+				"Failed to release CIDRs in replaced rules during policy add with replace. These may never be released.")
```
FWIW, the intent(*) is that the kvstore entry leasing system still provides protection around complete leakage of these things.

(*) Intent may or may not match reality. I don't believe we have tests specifically ensuring that these leases work as intended on the IPCache side.

raybejjani replied:
Right, but without a corresponding release the refcount is now always 1 (or more) too high for a specific CIDR, and any identity that relies on it. I can remove that last bit, though, to be less alarmist ;)
In this case, there is a refcount thing (https://github.com/cilium/cilium/blob/master/pkg/ipcache/kvstore.go#L296) that is separate from the refcount around the identity. My reasoning is that the first refcount will go wrong (the second one has some sync, maybe?). I might have followed this incorrectly, though, I tend to get lost in this part of the code :/

joestringer reviewed (approved):
New version LGTM, thanks Ray.

joestringer commented on daemon/policy.go:
```diff
@@ -206,27 +207,47 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
-			if len(tmp) > 0 {
-				d.dnsPoller.StopPollForDNSName(tmp)
+			oldRules := d.policy.SearchRLocked(opts.ReplaceWithLabels)
+			removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
```
Actually, coming back to this... Could this end up appending 'nil' to the list of removedPrefixes and end up in some weird situation later on?

raybejjani replied:
if `policy.GetCIDRPrefixes(oldRules)` returns a nil or an empty slice, then no. This is because the `...` appends each entry in the second parameter to the first (more akin to a python `.extend` vs `.append`). If the return includes a nil as an element, though, then is will append those elements as well. Is that a possibility? I can guard against it if you think it might come up.
Also, as I tested this (https://play.golang.org/p/vUA-9tBUyKm) I realised that we cannot append without the `...` because the type would be incorrect!

tgraf replied:
I think the code is fine, returning `[{ nil }]` would be a bug in `GetCIDRPrefixes()`

joestringer reviewed (dismissed):
One more thing to check up on (in multiple places). So long as this is resolved, I'm happy with the PR.
